### DISCUSSION TITLE
Is there a way to pull a statement date into renaming the document title?

### DISCUSSION THREAD

**Discussion by BlankSlate56** [jonaswinkler/paperless-ng#524] [Support]
I want to use paperless to scan in statements.  I was hoping it would be able to recognize the statement date and then be able to use that date when renaming the file.  for example, if I have a US Bank Statement from 2/1/2021, I want the name of the file to read "USB Statement 2-1-21.pdf".  I don't want to use the "added" or "created" date because those dates don't match the statement date.  Is this possible?

Also, I want to rename the files because I'll be pulling files into my bookkeeping software, and so it'll be much easier to find files by filename than going into paperless, finding the file, downloading it to a specific folder, and then loading it into the software.

TIA.

**Comment by jonaswinkler** [comment]
Hello, welcome to GitHub!

> I want to use paperless to scan in statements. I was hoping it would be able to recognize the statement date and then be able to use that date when renaming the file. for example, if I have a US Bank Statement from 2/1/2021, I want the name of the file to read "USB Statement 2-1-21.pdf". I don't want to use the "added" or "created" date because those dates don't match the statement date. Is this possible?

* Why exactly can't you use the created field? This field is intented to store the date at which a document was issued (invoice date, or issue date of the statement), which is not necessarily the date it was created in paperless.

* You can tell paperless to use the created field for naming files (see documentation, advanced topics)

> Also, I want to rename the files because I'll be pulling files into my bookkeeping software, and so it'll be much easier to find files by filename than going into paperless, finding the file, downloading it to a specific folder, and then loading it into the software.

See above. Manual renaming is not supported. However, you can tell paperless to derive filenames from correspondents / dates / types / etc.

**Reply by BlankSlate56** [reply]
Isn't created tied to when the PDF was scanned in?  My statement might have been from 2/1/21, I received it in the mail and scanned it in on 2/3/21, and added it to paperless on 2/5/21.  Am I misunderstanding the Created field?  Is that field automatically scanning for dates in the document?

**Reply by BlankSlate56** [reply]
I should add: per your last paragraph, I had read about creating the rules to rename per correspondents/date/type, etc, and that is what I was hoping to do.  My problem is that I don't believe the Created date is what I want in the file name.  If possible, I want paperless to pull my statement date off of the scanned pdf.

**Reply by shamoon** [reply]
Created is the date the document was created ie the invoice date, added is when you added it to paperless. Created is what you’re looking for I think

**Reply by BlankSlate56** [reply]
Then am I doing something wrong that Created is not matching my statement date, or is there a different "created" field that I'm not seeing?  Paperless pulled this document in and assigned the correspondent, document type, and tags correctly and automatically.  The created date does not match though.

Edit to add:  I scanned in this document on 1/22/21.  It was sitting in my filing cabinet and I'm working on digitizing all my records.

![image](https://user-images.githubusercontent.com/78878227/107654424-45d58b80-6c48-11eb-955d-b50483f5ebb6.png)

**Reply by shamoon** [reply]
Well yes sometimes it doesn’t correctly pick it up but that’s why it’s editable. See #450 and #188

**Reply by BlankSlate56** [reply]
Gotcha.  Now that I look at some other docs, the created date is actually from well before the statement date, and so there's no way that I would have scanned it in on that date.  The 3 or 4 I had looked at all had dates after the statement date, so I assumed those were the file creation dates (date they got scanned in)  This leads me to 2 more questions:

1.  If I manually edit the Created field, does paperless learn from that edit?  Meaning, will it search the document for a matching date so that it will start interpreting dates correctly for that correspondent/file type?
2. If the created date is not pulled in correctly, is there any way to re-run the file naming rule once the field has been corrected so that the file is stored with the correct name for access outside of paperless?

Thanks for all the help, I'm really interested to see where I can take this.  I've been scanning in hundreds of docs and renaming them all manually.  If I can figure this out, it'll really streamline my workflow.

**Reply by jonaswinkler** [reply]
> Gotcha. Now that I look at some other docs, the created date is actually from well before the statement date, and so there's no way that I would have scanned it in on that date. The 3 or 4 I had looked at all had dates after the statement date, so I assumed those were the file creation dates (date they got scanned in) 

Paperless looks for dates in the OCR'ed content and assigns whatever it finds first to the "created" field. This works reasonably well for letters, sind the date the letter was sent (e.g., *created*) usually appears in its header. However, it does not always work. Sometimes there are other things before the actual date that look like dates, and paperless uses that. Sometimes OCR produces garbage and paperless won't recognize the date.

> This leads me to 2 more questions:

> 1. If I manually edit the Created field, does paperless learn from that edit?  Meaning, will it search the document for a matching date so that it will start interpreting dates correctly for that correspondent/file type?

No.

> 2. If the created date is not pulled in correctly, is there any way to re-run the file naming rule once the field has been corrected so that the file is stored with the correct name for access outside of paperless?

Everytime you save a document, paperless will update the filename of that document according to `PAPERLESS_FILENAME_FORMAT`. It will also create and move folders as necessary.

> 
> Thanks for all the help, I'm really interested to see where I can take this. I've been scanning in hundreds of docs and renaming them all manually. If I can figure this out, it'll really streamline my workflow.

**Reply by BlankSlate56** [reply]
I've run about 20 of those statements through now, and paperless didn't match the Created date correctly on a single one.  It has tagged everything else correctly though, so I'm excited about that.  If all I have to do is update the Created box and it'll update the file name on save, that's still a massive improvement over my current system.

Would it even be possible to tie the Created field into the machine learning so that it could start learning where to find the correct date?  When I look at the OCR data, the date is usually there within the first couple of lines, but paperless isn't finding it.   If it could be trained somehow it might help.  I could submit a feature request if it is possible, but won't bother if it isn't.

Thanks again for your work, this is starting to look like it could be very useful for me.

**Reply by jonaswinkler** [reply]
> I've run about 20 of those statements through now, and paperless didn't match the Created date correctly on a single one.

That's probably because all those look about the same and paperless fails on every one for the same reason.

If you could share the exact content up until after the date, that would be great. Of course, replace any confidential data with similar looking bogus data (i.e., numbers with numbers, letters with letters)

> Would it even be possible to tie the Created field into the machine learning so that it could start learning where to find the correct date?

This requires a different approach / algorithm than what is used for the Auto matching algorithm. The appropriate algorithms also require a lot more computing resources and much more training data, so it will be difficult to get this running reliably.

**Reply by BlankSlate56** [reply]
When you were asking for content, I assume you were asking for copies of the OCR data?  If so, below are 3 samples.  Thanks for all your help.  I might toss a feature request out to improve the Created data algorithm just to gauge interest, but understand if it isn't a priority.  I'm still working my way through all this, but it appears to be a huge improvement over what I currently do, so thanks again.

As you can see in the examples, the date is almost always one of the first things in the data.  This is pretty consistent through the other 20 files too.  I haven't tried any statements from other correspondents yet; that's part of my project for today.

Example 1:
YNNN
05/25/2020)
11230900
eek e ¢ rt ee oF) ©
Cena
1111 111th Ave. * PO Box 8 » Narnia, XX 11111

Example 2:
06/25/2020
11230900
B [= Cc T A 1 Cc
Eas
1111 111th Ave. * PO Box 8 » Narnia, XX 11111

Example 3:
01/25/2020
11230900
§
1111 111th Ave. * PO Box 8 » Narnia, XX 11111

**Reply by BlankSlate56** [reply]
Also, just to clarify:  The Created field is not editing any metadata, correct?  When I go through File Explorer and look at a file, the Date Created there does not match the Created date in paperless.  The date created in the metadata appears to be the date the paperless consumed the original and created the new file.

**Reply by BlankSlate56** [reply]
As I work with this some more and start to understand the workflow and terminology better, I realize I probably caused some confusion by referring to the document title in my subject line when I actually meant the document filename.  However, this does lead me to another question that I think  you might have answered already:  Can the Title be automatically generated by paperless similar to the filename renaming rules?  Is there a reason why I'd want the title to be separate from the filename?

I didn't see anything in the docs talking about automatic titles, but I just wanted to make sure.

**Reply by jonaswinkler** [reply]
> When you were asking for content, I assume you were asking for copies of the OCR data?

Exactly, the text that's displayed in the content field. Thank you, I'll see if I can work out what's going here, this looks like it doesn't do the expected thing.

> The Created field is not editing any metadata, correct?

Yes. The content of the files is never modified after they're added to paperless. However, I could probably change the created date of the file system attributes according to what is set in paperless.



> I realize I probably caused some confusion by referring to the document title

No worries. I've got what you were trying to say.



> Can the Title be automatically generated by paperless similar to the filename renaming rules?

This is a complex topic and many users use the title differently. On invoices/receipts, I like to put the name of the product I bought in the title so I can easily identify that document if I need it for warranty reasons. I know that there are algorithms out there that are able to pull that kind of information from text data, but they usually require a lot more data than what the average paperless user has stored in the database right now.

> Is there a reason why I'd want the title to be separate from the filename?

I'd like the title of documents to be short and descriptive, and since paperless also stores correspondents and types separately, there's no reason to include that kind of information in the title. However, I'd like my filenames to be somewhat more descriptive in case I need to find something there.

**Reply by BlankSlate56** [reply]
Just an fyi, those examples I posted are only partial datasets - basically the first 5-6 lines of data.  I didn't post the entire dataset to avoid accidentally missing any identifying data.  If you want the entire dataset I can work on getting it for you.

> Yes. The content of the files is never modified after they're added to paperless. However, I could probably change the created date of the file system attributes according to what is set in paperless.
>

Personally, I'd like to see this, but the Created field would have to be more effective than it is now, or you'd have to allow paperless to edit metadata via the Created field in the UI if I need to correct a date that was pulled incorrectly.

> This is a complex topic and many users use the title differently. On invoices/receipts, I like to put the name of the product I bought in the title so I can easily identify that document if I need it for warranty reasons. I know that there are algorithms out there that are able to pull that kind of information from text data, but they usually require a lot more data than what the average paperless user has stored in the database right now.
> 

Personally, many of my invoices either are for the same product that I'm buying repeatedly or a single invoice covers tons of products that would be too lengthy to list.  Either way, having it describe the product instead of the document is either repetitive to too lengthy.  I'd prefer to have the ability have rules to format the title in the same way we can format the filename.

> I'd like the title of documents to be short and descriptive, and since paperless also stores correspondents and types separately, there's no reason to include that kind of information in the title. However, I'd like my filenames to be somewhat more descriptive in case I need to find something there.

I was already manually creating filenames before and am trying to use this app to reduce manual data entry.  If I have to manually enter a title each time, then it kind of defeats the purpose to a certain extent.  I realize all use cases are different, so I'm just trying to better understand how this was developed to be used, and how I might be able to adapt it to me or me to it.

Thanks again.

**Reply by BlankSlate56** [reply]
In the vein of "adapting me to it", I have discovered a use for Title that while it isn't probably the intended use, it does work for me.  I have created the following filename format: {correspondent}/{correspondent} {document_type} {title} {created}

This obviously creates a folder for each correspondent, and then gives a filename something along the lines of "Megastore Contract <title> 2021-01-01".  I am using <title> to manually enter unique contract numbers into the filename within the paperless UI.  Suppose I have 2 Megastore Contracts from 1/1/2021:  1112 & 1113.  By manually entering the contracts into the title, my files now get renamed as "Megastore Contract 1112 2021-01-01" & "Megastore Contract 1113 2021-01-01".  This allows me to know what the file is, even though paperless isn't actually pulling that data into the filename rule automatically.  If a document doesn't have a number that needs to be identified in the filename, then I just leave Title blank.  All that happens then is that an extra space shows up in the filename that you don't even notice.

It'd be slick if we could have trained filename rules that'd learn to look for a contract number within the context of other data (for example, teach it that the 4 numerals following the word "contract" for a Megastore document are the contract number), but I realize that's probably a very complicated chunk of code.

**Reply by BlankSlate56** [reply]
> I'll see if I can work out what's going here, this looks like it doesn't do the expected thing.

@jonaswinkler if you are still looking into this, I've loaded about 160 files in the last couple of days.  These files cover 19 different correspondents and 13 different document types, so it isn't like I'm force-feeding it nearly identical copies of the same invoices.  "Created" hasn't pulled the correct date for a single one of them, despite the dates being near the top of the documents probably 90% of the time.  I don't mean this as a complaint so much as a report.  

I have noticed that maybe 25% of the time, the date that is pulled into "created" is exactly 1 day ahead of the statement/letter/invoice date that I want.  Meaning, if the invoice date is 2021/01/01, paperless is pulling the date in as 2020/12/31.  This happened frequently enough that it isn't just a coincidence.  Perhaps there is something broken in the Create date algorithm?  I'm running 1.1.0.

**Reply by BlankSlate56** [reply]
For what its worth, below is an example of one of the datasets where the created date was pulled in 1 day early.  paperless read this as June 8 2020 instead of June 9.  Had to make quite a few changes to parse out identifiable data, don't know how much that'll affect your testing.

North Zeeland Family Dentistry
1400 notaroad Drive
South Nantucketato, NM 50630
(057)265-2237 (570)652-2723
June 09, 2020
Account Aging
Current: $0.00
Fred Flintstone 30 Days: $0.00
19604 570th Ave 60 Days: $0.00
Bells, NM 51607 90 Days: $0.00

**Reply by shamoon** [reply]
I think the 1 day thing is actually a time zone issue because it’s stored in UTC, I’ve seen it before. We still have some work to do on displaying / entering dates in the UI which I think will help. 

And agreed the date matching can be inaccurate often, not sure if thats changed from original paperless or not

**Reply by jonaswinkler** [reply]
> Meaning, if the invoice date is 2021/01/01, paperless is pulling the date in as 2020/12/31. This happened frequently enough that it isn't just a coincidence. Perhaps there is something broken in the Create date algorithm? I'm running 1.1.0.

Set your time zone in the configuration files correctly. The default is UTC. Therefore, any dates assigned to new documents on a certain date at midnight (that's what paperless uses) will display one day early when you're in a time zone that's behind UTC (US, for example).

This change is not retroactive. However, configuring your time zone correctly will cause any new documents to receive the correct date.

I don't exactly know why they added time zone support in OG paperless, but it's there, and I haven't changed it for compatibility reasons.

**Reply by BlankSlate56** [reply]
ok, I'll check out the time zone thing.  I know I haven't tweaked that so I probably missed it in initial setup.  That'd make sense for why some of the pulls were one day off.  Thanks again.

I actually thought manually entering the dates was easier than I was expecting, I was just hoping that I wouldn't have to do so much of it.  Its still fun watching the AI learn the auto-tags and apply them correctly over a growing range of docs.

**Comment by balajeek** [comment]
Hi, thanks for the awesome project.  I am in the process of setting it up, did some few tests to get a feel of it, then i had the same question as this thread subject says:

I have been doing my document management (somewhat manually) and i use Fujitsu scanner which scans and names the files based on what i highlight on the statement i am scanning.   There is also a tool hazel on mac were we can right rules so it will find the dates on a scanned document and name them and move them to the approprite folder.  I can continue to do that.

But is there a way to tell paperless to take {created_year} {created_month} from the file name that it consumes rather than form the ocr content, because seeing from this long thread mentions paperless most often does not pick the right date from the document, and i tested a dozen document from archive and all are many years off, a 2012 bank of america statement goes into 2017 folder.

possible?
